File: C++/4.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        /*
            这个题是我乱改出来的.....
        */
        int n = nums1.size(), m = nums2.size();
        int k = (n + m + 1) / 2;
        vector<int>::iterator begin1 = nums1.begin(), end1 = nums1.end();
        vector<int>::iterator begin2 = nums2.begin(), end2 = nums2.end();
        vector<int>::iterator mid1;

        /*
            首先二分nums1，假设当前位置为mid，在nums2中查找nums1[mid]的upper_bound，
            将两个数组分成大于等于nums1[mid]和小于等于nums1[mid]两部分，
            直到出现下述情况：
            1. 该部分为空；
        */
        while(begin1 < end1)
        {
            mid1 = begin1 + (end1 - begin1) / 2;
            auto it2 = upper_bound(begin2, end2, *mid1);
            int prev_size = (mid1 - begin1) + (it2 - begin2);
            if(k > prev_size) k -= prev_size, begin1 = mid1, begin2 = it2;
            else end1 = mid1, end2 = it2;
            /*
            2. 说不清楚的情况。
            */
            if(end1 - begin1 == 1 && (end2 - begin2 == 0 || *begin2 > *begin1)) break;
        }
        // tag 表示中间的数在num1还是nums2
        int tag;
        if(end1 - begin1 == 1)
        {
            if(k == 1) tag = 1, mid1 = begin1;
            else tag = 2, mid1 = begin2 + k - 2;
        }
        else tag = 2, mid1 = begin2 + k - 1;
        if((n + m) % 2 != 0) return static_cast<double>(*mid1);
        else
        {
            double sum = *mid1;
            vector<int>::iterator mid2;
            if(tag == 1)
            {
                mid2 = begin2;
                if(mid2 == nums2.end()) sum += *(mid1 + 1);
                else if(mid1 + 1 == nums1.end()) sum += *mid2;
                else sum += min(*(mid1 + 1), *mid2);
            }
            else
            {
                mid2 = lower_bound(end1, nums1.end(), *mid1);
                if(mid2 == nums1.end()) sum += *(mid1 + 1);
                else if(mid1 + 1 == nums2.end()) sum += *mid2;
                else sum += min(*(mid1 + 1), *mid2);
            }
            return sum / 2.0;
        }
        // 总之很乱
    }
};

File: Python-jk/4.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        l1=len(nums1)
        l2=len(nums2)
        L=l1+l2
        a=nums1
        b=nums2
        i=0
        j=0
        tmp=0
        tmp1=0
        tmp2=0
        if L%2:
            while (i+j)<L/2:
                if (i<l1)&(j<l2):
                    tmp=min(a[i],b[j])
                    if (i+1<l1)&(j+1<l2):
                        if (a[i+1]<b[j+1]):
                            i+=1
                        else:
                            j+=1
                    elif i+1==l1:
                        if (a[i]==tmp):
                            i+=1
                        else:
                            j+=1
                    elif j+1==l2:
                        if (b[j]==tmp):
                            j+=1
                        else:
                            i+=1
                elif i==l1:
                    tmp=b[j]
                    j+=1
                elif j==l2:
                    tmp=a[i]
                    i+=1
            return tmp
        else:
            tmp=min(a[0],b[0])
            if tmp==a[0]:
                i+=1
            else:
                j+=1
            while True:
                if (i+j)>L/2:
                    break
                if (i<l1)&(j<l2):
                    tmp1=min(a[i],b[j])
                    if (i+1<l1)&(j+1<l2):
                        if (a[i+1]<b[j+1]):
                            i+=1
                        else:
                            j+=1
                    elif i+1==l1:
                        if (a[i]==tmp1):
                            i+=1
                        else:
                            j+=1
                    elif j+1==l2:
                        if (b[j]==tmp1):
                            j+=1
                        else:
                            i+=1
                elif i==l1:
                    tmp1=b[j]
                    j+=1
                elif j==l2:
                    tmp1=a[i]
                    i+=1
                if (i<l1)&(j<l2):
                    tmp2=min(a[i],b[j])
                    if (i+1<l1)&(j+1<l2):
                        if (a[i+1]<b[j+1]):
                            i+=1
                        else:
                            j+=1
                    elif i+1==l1:
                        if (a[i]==tmp2):
                            i+=1
                        else:
                            j+=1
                    elif j+1==l2:
                        if (b[j]==tmp2):
                            j+=1
                        else:
                            i+=1
                elif i==l1:
                    tmp2=b[j]
                    j+=1
                elif j==l2:
                    tmp2=a[i]
                    i+=1
            return (tmp1+tmp2)/2

            
a=Solution()
b=a.findMedianSortedArrays([1,3],[2])
print(b)
b=a.findMedianSortedArrays([1],[2,3])
print(b)
b=a.findMedianSortedArrays([3],[1,2])
print(b)